os/20211014-s390x-secure-execution.md

```diff
@@ -0,0 +1,241 @@
+
+## Step 2: New `qemu-secex.qcow2` build artifact for CoreOS with enabled Secure Execution.
+
+`coreos-xyz-qemu-secex.qcow2` image with `dm-verity` setup for `/boot` and `rootfs` is available for download. This is done by using new `buildextend-secex` `coreos-assembler` command, which:
+ - adds new `/se` partition for `sdboot`
+ - adds `roothash` and `boothash` partitions for [dm-verity](https://docs.kernel.org/admin-guide/device-mapper/verity.html) setup
+ - calculates hashes for `/boot` and `/` and appends them to kernel's cmdline
+ - generates `/se/sdboot` encrypted by universal public hostkey (valid for all Z15 and newer machines)
+ - `zipl`es `/se/sdboot`, so `qemu-secex.qcow2` image could only be run with Secure Execution
+
+Customer should add his public hostkeys(s) into `ignition` config.
+
+## Step 3: During firstboot `/boot` and `rootfs` are encrypted with randomly generated LUKS keys; `integrity` option is enabled
+
+During firstboot `/boot` and `/` are luks encrypted with randomly generated luks keys. This ensures, that each CoreOS isntance has its own LUKS keys and also makes it not possible to read the contents of qcow2 image just by inspecting (`libguestfish`) it. Any modifications of qcow image will be detected in runtime by [dm-integrity](https://docs.kernel.org/admin-guide/device-mapper/dm-integrity.html).
+
+
+After provisioning, new disk is like:
+```
+NAME         PARTLABEL TYPE  MOUNTPOINT
+vdb                    disk
+|-vdb1       se        part
+|-vdb3       boot      part
+| `-boot_dif           crypt
+|   `-boot             crypt /boot
+`-vdb4       root      part
+  `-root_dif           crypt
+    `-root             crypt /sysroot
+
+```
+
+- `se`   - partition for newly generated`sdboot` encrypted image
+- `boot` - LUKS encrypted boot partition
+- `root` - LUKS encrypted root partition
+- `boot_dif` - `dm-integrity` for `/boot`
+- `root_dif` - `dm-integrity` for `/root`
+
+## Step 4: Ignition protection
+
+Proposal:
+
+The user will encrypt the Ignition config with his symmetric key and the Ignition public key for the pre-built image.
+The private key is part of the `sdboot` image and is randomly generated during qcow2 image build. The public key is part of the build metadata. During firstboot the Ignition config gets decrypted.
+We should avoid printing the Ignition config to the console (e.g. serial and VGA) and scrub it before any shell access is possible.
+
+- In the SE path, add dropins to disable Ignition logging
+- Right after config decryption drop the ignition's private key to avoid leaking it.
+- In the emergency shell code, delete `/run/ignition.json` before entering the shell.
+- User generates the symmetric `key` and encrypts with it the `config.ign`:
+  ```
+  openssl rand -base64 96 > key
+  openssl enc -salt -pbkdf2 -aes256 -in config.ign -out config.crypted -pass file:./key
```

Comment:
It's awkward and error-prone to manually perform low-level operations such as generating the symmetric key.  We should provide tooling for this, or preferably use existing tools.  Is GPG a reasonable alternative?  The key management would be sort of awkward, though; the user would end up with a number of public keys for random OS images on their keyring.

Reply:
ok, GPG may make life easier here. i'll update this and WIP-PRs